File: frontend/src/app/services/lobby.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';
import { Room, CreateRoomPayload } from '../core/models/room.model';

@Injectable({ providedIn: 'root' })
export class LobbyService {
  private readonly API = `${environment.apiUrl}/lobby`;

  constructor(private http: HttpClient) {}

  getRooms(): Observable<{ rooms: Room[] }> {
    return this.http.get<{ rooms: Room[] }>(`${this.API}/rooms`);
  }

  createRoom(payload: CreateRoomPayload): Observable<{ room: Room }> {
    return this.http.post<{ room: Room }>(`${this.API}/rooms`, payload);
  }

  joinRoom(code: string): Observable<{ room: Room }> {
    return this.http.post<{ room: Room }>(`${this.API}/rooms/join`, { code });
  }

  getRoom(code: string): Observable<{ room: Room }> {
    return this.http.get<{ room: Room }>(`${this.API}/rooms/${code}`);
  }

  getLeaderboard(): Observable<{ leaderboard: any[] }> {
    return this.http.get<{ leaderboard: any[] }>(`${this.API}/leaderboard`);
  }
}

File: queryquest_app/lib/services/lobby_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';
import '../models/room.dart';
import 'api_service.dart';

class LobbyService {
  final String _apiUrl = '${ApiConfig.baseUrl}/lobby';

  Future<Map<String, String>> _getHeaders() async {
    final prefs = await SharedPreferences.getInstance();
    final token = prefs.getString('qq_token');
    return {
      'Content-Type': 'application/json',
      if (token != null) 'Authorization': 'Bearer $token',
    };
  }

  Future<List<Room>> getRooms() async {
    final response = await http.get(
      Uri.parse('$_apiUrl/rooms'),
      headers: await _getHeaders(),
    );
    if (response.statusCode == 200) {
      final data = jsonDecode(response.body);
      return (data['rooms'] as List).map((r) => Room.fromJson(r)).toList();
    } else {
      throw Exception('Failed to load rooms');
    }
  }

  Future<Room> createRoom(Map<String, dynamic> payload) async {
    final response = await http.post(
      Uri.parse('$_apiUrl/rooms'),
      headers: await _getHeaders(),
      body: jsonEncode(payload),
    );
    if (response.statusCode == 200 || response.statusCode == 201) {
      final data = jsonDecode(response.body);
      return Room.fromJson(data['room']);
    } else {
      throw Exception('Failed to create room');
    }
  }

  Future<Room> joinRoom(String code) async {
    final response = await http.post(
      Uri.parse('$_apiUrl/rooms/join'),
      headers: await _getHeaders(),
      body: jsonEncode({'code': code}),
    );
    if (response.statusCode == 200) {
      final data = jsonDecode(response.body);
      return Room.fromJson(data['room']);
    } else {
      throw Exception('Failed to join room');
    }
  }

  Future<List<dynamic>> getLeaderboard() async {
    final response = await http.get(
      Uri.parse('$_apiUrl/leaderboard'),
      headers: await _getHeaders(),
    );
    if (response.statusCode == 200) {
      final data = jsonDecode(response.body);
      return data['leaderboard'] as List<dynamic>;
    } else {
      throw Exception('Failed to load leaderboard');
    }
  }
}
